### PR TITLE
Support running a suite of tests concurrently

### PR DESCRIPTION
**Pull request by lmars**
ping @titanous

### CODE REVIEW

File: run.go [hunk 1]
```diff
@@ -12,13 +12,25 @@ import (
...
 // Suite registers the given value as a test suite to be run. Any methods
 // starting with the Test prefix in the given value will be considered as
 // a test method.
 func Suite(suite interface{}) interface{} {
-	allSuites = append(allSuites, suite)
+	allSuites = append(allSuites, s{suite, false})
+	return suite
+}
+
+var ConcurrencyLevel = 5
```
Comment (titanous) [inline run.go]: I think this would make more sense in the `RunConf`.

File: check.go [hunk 2]
```diff
@@ -602,11 +604,29 @@ func (runner *suiteRunner) run() *Result {
...
-					c := runner.runTest(runner.tests[i])
-					if c.status == fixturePanickedSt {
-						runner.skipTests(missedSt, runner.tests[i+1:])
-						break
+				if runner.concurrent {
+					queue := make(chan struct{}, ConcurrencyLevel)
+					for i := 0; i < ConcurrencyLevel; i++ {
+						queue <- struct{}{}
+					}
+					for i := 0; i != len(runner.tests); i++ {
```
Comment (titanous) [inline check.go]: `i < len(runner.tests)`

Comment (titanous) [inline check.go]: Can we just do:

``` go
for i := range runner.tests {
```

Comment (titanous) [inline check.go]: or even:

``` go
for _, t := range runner.tests {
```

File: check.go [hunk 3]
```diff
@@ -602,11 +604,29 @@ func (runner *suiteRunner) run() *Result {
...
+						<-queue
+						go func(i int) {
+							runner.runTest(runner.tests[i])
+							queue <- struct{}{}
+						}(i)
+					}
+					for i := 0; i < ConcurrencyLevel; i++ {
+						<-queue
+					}
+					close(queue)
```
Comment (titanous) [inline check.go]: Probably no reason to do this.

File: check.go [hunk 4]
```diff
@@ -602,11 +604,29 @@ func (runner *suiteRunner) run() *Result {
 		if runner.checkFixtureArgs() {
 			c := runner.runFixture(runner.setUpSuite, "", nil)
 			if c == nil || c.status == succeededSt {
-				for i := 0; i != len(runner.tests); i++ {
-					c := runner.runTest(runner.tests[i])
-					if c.status == fixturePanickedSt {
-						runner.skipTests(missedSt, runner.tests[i+1:])
-						break
+				if runner.concurrent {
+					queue := make(chan struct{}, ConcurrencyLevel)
```
Comment (titanous) [inline check.go]: Call this `bucket`, as it's a variant of the token bucket data structure.

File: check.go [hunk 5]
```diff
@@ -602,11 +604,29 @@ func (runner *suiteRunner) run() *Result {
...
+							runner.runTest(runner.tests[i])
+							queue <- struct{}{}
+						}(i)
+					}
+					for i := 0; i < ConcurrencyLevel; i++ {
+						<-queue
+					}
+					close(queue)
+				} else {
+					for i := 0; i != len(runner.tests); i++ {
```
Comment (titanous) [inline check.go]:
``` go
for i := range runner.tests {
```

**Comment by lmars** [comment]
@titanous comments addressed

File: check.go [hunk 6]
```diff
@@ -544,14 +547,16 @@ func newSuiteRunner(suite interface{}, runConf *RunConf) *suiteRunner {
...
+		suite:            suite,
+		output:           newOutputWriter(conf.Output, conf.Stream, conf.Verbose),
+		tracker:          newResultTracker(),
+		benchTime:        conf.BenchmarkTime,
+		benchMem:         conf.BenchmarkMem,
+		tempDir:          &tempDir{},
+		keepDir:          conf.KeepWorkDir,
+		tests:            make([]*methodType, 0, suiteNumMethods),
+		concurrent:       concurrent,
+		concurrencyLevel: conf.ConcurrencyLevel,
```
Comment (titanous) [inline check.go]: If concurrent is set, this needs to be greater than zero (default it to 1?).

**Comment by lmars** [comment]
@titanous added support for running the suites concurrently: 9f34167

Going to look at the concurrent writes now to see if there is an issue.

**Comment by lmars** [comment]
@titanous comments addressed, the logging all looks to be thread-safe.

**Comment by titanous** [comment]
Awesome! LGTM.

**Pull request merged**